os/pm/pm_timer/pm_timer_delete.c

```diff
@@ -0,0 +1,126 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/pm/pm.h>
+#include <tinyara/irq.h>
+
+#include "pm_timer.h"
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Public Variables
+ ************************************************************************/
+
+/************************************************************************
+ * Private Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Name: pm_timer_delete
+ *
+ * Description:
+ *   This function deletes a pm timer. It either returns it to the 
+ *   g_pmTimer_freeList or frees it if it was created dynamically.
+ *
+ * Parameters:
+ *   pm timer to be deleted
+ *
+ * Returned Value:
+ *   None.
+ *
+ ************************************************************************/
+
+void pm_timer_delete(pm_wakeup_timer_t *timer) 
+{
+        irqstate_t state;
+
+        DEBUGASSERT(timer);
+
+        /* Check if this needs to be atomic */
+        state = enter_critical_section();
+
+        if (PM_ISSTATIC(timer)) {
+                /* Put the pm timer back on the free list */
+                sq_addlast((FAR sq_entry_t *)timer, &g_pmTimer_freeList);
+                g_pmTimer_nfree++;
+                DEBUGASSERT(g_pmTimer_nfree <= CONFIG_PM_MAX_STATIC_TIMER);
+                leave_critical_section(state);
+
+        } else if (PM_ISALLOCED(timer)) {
+                /* Free the dynamically allocated timer 
+                * If the timer was released from an interrupt handler, 
+                * sched_kfree() will defer the actual deallocation of the
+                * memory until a more appropriate time. 
+                * 
+                * We don't need interrupts disabled to do this */
+                leave_critical_section(state);
+                sched_kfree(timer);	
+
+        } else {
+                /* This should not be the case because a pm timer 
+                * is either statically allocated or dynamically */
+                pmdbg("This should not be any case\n");
+                leave_critical_section(state);
+        }
```

Comment:
All condition here does leave_critical_section, how about pull it out and do it once at the end of pm_timer_delete?